### PR TITLE
MRG: Initial support for NIRS data

### PR DESCRIPTION
**Pull request by rob-luke**
PR Description
--------------

This PR adds support for fNIRS. There is a BEP in progress for NIRS support here... https://github.com/bids-standard/bids-specification/issues/438

**This is a work in progress of a work in progress document. It should not be relied upon for storage of data.**

I am maintaining this PR to stay up to date with the BEP. The plan is to use this PR to test what works and doesn't in the document, and iterate accordingly.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

### CODE REVIEW

**Comment by agramfort** [comment]
@rob-luke is the bids validator working already with fNIRS? if not I would start there

**Comment by rob-luke** [comment]
No it's not. Im working on that now too. Will focus my efforts there first then.

**Comment by sappelhoff** [comment]
cross link: https://github.com/bids-standard/bids-validator/pull/952

:+1: good idea to open PRs for this early @rob-luke

**Comment by jasmainak** [comment]
You will need the validator for testing in `mne-bids`

**Comment by sappelhoff** [comment]
> You will need the validator for testing in mne-bids

but you can use your own bids-validator fork for testing in this mne-bids PR ... that way you get your own experimental bids-validator nirs support for the changes you want to introduce here.

You'd have to change this line:

https://github.com/mne-tools/mne-bids/blob/1ce1717f923843fc5eea970c6aa44e95c018bef0/.travis.yml#L38-L39

and this one, to point to your fork:

https://github.com/mne-tools/mne-bids/blob/1ce1717f923843fc5eea970c6aa44e95c018bef0/.travis.yml#L69

and similarly in Appveyor.

**Comment by rob-luke** [comment]
> But of course we can only merge here, once your validator PR is merged. (So this PR will probably also have to be rebased regularly)

Absolutely. Thanks



> you can use your own bids-validator fork for testing in this mne-bids PR

This is some dark git magic. Love it, thanks for the pointer.

**Comment by rob-luke** [comment]
Ok I will get back to this soon, and looks like I will need to rebase. I think most of the behind the scenes work for snirf is now completed at https://github.com/mne-tools/mne-python/pull/8079 and https://github.com/mne-tools/mne-nirs/pull/151 and https://github.com/mne-tools/mne-python/pull/7972

**Comment by jasmainak** [comment]
sounds good!

**Comment by larsoner** [comment]
@rob-luke can we move forward with this somehow? One option is that when reading fNIRS files we emit a warning that it's experimental. I think it "just" needs a rebase or -- maybe easier -- merge with `main`.

**Comment by rob-luke** [comment]
Getting this merged will save me a headache and simplify the tutorials in MNE-NIRS too!!  That would be fantastic if we can merge this before the spec is accepted (from early comments above I got the impression this was not acceptable). I will rebase then you can add whatever warnings you'd expect.

**Comment by larsoner** [comment]
> That would be fantastic if we can merge this before the spec is accepted (from early comments above I got the impression this was not acceptable)

It looks like this PR has been open for a year and a half, any additional arguments you can make in its favor based on that to persuade folks? Like it's needed to read data from some studies with public data in the wild, has been pretty stable, etc.? At some point it seems like merging something experimental might be worth it if the spec expansion is moving very slowly, but not expected to deviate much from what this PR does.

But I am far from a BIDS expert, so I might be way off on this...

**Comment by rob-luke** [comment]
I think we want to get this merged because the spec is stable (we are just writing the manuscript now), has been implemented as is in the other toolboxes (Homer), is used for several publicly available studies.

But I am having a hard time rebasing. I might just make a new branch from main and copy the changes. Its a mess for such old PRs :( Unless you have another tip?

**Comment by larsoner** [comment]
Rebasing 43 commits is a nightmare. I would `git rebase -i HEAD~43`, squash all but one commit, *then* `git fetch upstream && rebase upstream/main`. Then you only have to fight with one set of file conflicts, not one set of file conflicts per (worst case) each of the 43 commits.

**Comment by larsoner** [comment]
... but sometimes copy-paste is easy enough. Since this is just `-16`, it seems like that would hopefully be pretty easy as you're not modifying much existing code.

Either way, though, you could force-push here rather than opening a separate PR. For example:
```
git checkout nirs
git reset --hard origin/nirs
git push origin nirs:nirs-bak
git reset --hard upstream/main
... # do your copy-paste-commit
git push origin --force nirs
```
Then this PR just sees your one new commit, but preserves the conversation.

**Comment by rob-luke** [comment]
I started adding some tests, but we will need to wait for https://github.com/mne-tools/mne-testing-data/pull/91 to be merged first.

**Comment by rob-luke** [comment]
Great! The first test is passing on `mne-main, validator-main (my branch)` with the new testing data.
I will add a few more tests and look through the code myself (2020 was a long time ago 😱 ). Then I will ping when ready for more eyes

**Comment by larsoner** [comment]
@rob-luke the easy way out here might be to `@pytest.mark.skipif(not check_version('mne', '1.0')` the tests and in the code `if not check_version('mne', '1.0'): raise RuntimeError('fNIRS support in MNE-BIDS requires MNE-Python version 1.0')` or so. Then this code will only live for a brief time in the upcoming (whenever it happens) MNE-BIDS release, and gets removed from `main` afterward.

**Comment by hoechenberger** [comment]
@rob-luke I've tried to implement @larsoner's changes, didn't run the test suite locally though. Let's see if it starts passing now…

File: mne_bids/tests/test_write.py [hunk 1]
```diff
@@ -1587,6 +1589,41 @@ def test_eegieeg(dir_name, fname, reader, _bids_validate, tmp_path):
         _bids_validate(output_path)
 
 
+@pytest.mark.skipif(not check_version('mne', '1.0'))
```
Comment (larsoner) [inline mne_bids/tests/test_write.py:1592]:
```suggestion
@pytest.mark.skipif(not check_version('mne', '1.0'), reason='Needs MNE 1.0')
```

Comment (hoechenberger) [reply mne_bids/tests/test_write.py]: Already addressed just seconds before your comment :)

**Comment by larsoner** [comment]
Failure is just coverage. @rob-luke ready to go from your end? This is still in draft mode and WIP

**Comment by hoechenberger** [comment]
@larsoner I'm not sure if this code ever received a proper code review from any of the other MNE-BIDS devs. I can do that tomorrow. But we should definitely try to get test coverage up in any case.

**Comment by hoechenberger** [comment]
All green now. But I think we're still missing a few lines that should be covered by tests.

File: .github/workflows/unit_tests.yml [hunk 2]
```diff
@@ -175,7 +175,7 @@ jobs:
         pushd ..
         # Get npm 7+
         npm install -g npm
-        git clone --depth 1 https://github.com/bids-standard/bids-validator
+        git clone --depth 1 --branch nirs https://github.com/rob-luke/bids-validator.git
```
Comment (rob-luke) [inline .github/workflows/unit_tests.yml:178]: Revert this before merge!

**Comment by rob-luke** [comment]
Thanks @larsoner and @hoechenberger ! 

> I'm not sure if this code ever received a proper code review from any of the other MNE-BIDS devs.

Correct. This hasn't received a full code review, maybe there are nicer ways than what I implemented. Maybe @sappelhoff would like a chance to review?

Also, I would add that this adds minimum required support for the BIDS NIRS spec. There are some optional fields that I have not implemented yet. But I can add the MAY/OPTIONAL fields in future PRs.

**Comment by rob-luke** [comment]
I have not updated the docs, as the specification has not been officially submitted or accepted in to BIDS yet. Or would you prefer me to update the docs pages to include fNIRS in the introduction etc?

**Comment by larsoner** [comment]
> There are some optional fields that I have not implemented yet. But I can add the MAY/OPTIONAL fields in future PRs... I have not updated the docs, as the specification has not been officially submitted or accepted in to BIDS yet. Or would you prefer me to update the docs pages to include fNIRS in the introduction etc?

I am no (MNE-)BIDS expert, but if you can work with fNIRS data in MNE-BIDS, I would expect an entry for fNIRS in the docs. Ideally, since it's (still to some degree at least) experimental and incomplete, I'd put first thing after the `fNIRS` heading a `.. warning::` about it being experimental and incomplete, and subject to potential incompatibility with the finalized standard, change without any deprecation, etc. (even though we hope / expect that what is here will be correct in the end, and support some/most important existing use cases!).

> Revert this before merge!

I see this above about `bids-validator` -- so does that mean https://github.com/bids-standard/bids-validator/pull/952 needs to be merged first? It looks like it might also be close so maybe that's doable...

**Comment by rob-luke** [comment]
> does that mean https://github.com/bids-standard/bids-validator/pull/952 needs to be merged first? It looks like it might also be close so maybe that's doable...

I am not 100% sure of the order of events here. I don't know if the specification gets accepted, then the changes to the validator. Or if the validator gets merged, then the spec is submitted etc. Or if everything effectively happens at once. The specification submission is almost ready, we are just in the second round of reviews on the manuscript and getting a dataset prepared from Donders/Fieldtrip to complement the MNE created dataset. But this could take a month or so before we hit the submit button. In short, the PR may be ready, but may not be accepted until all these other bits fall in to place. @sappelhoff would know best how this works.

**Comment by sappelhoff** [comment]
> I see this above about bids-validator -- so does that mean https://github.com/bids-standard/bids-validator/pull/952 needs to be merged first? It looks like it might also be close so maybe that's doable...

it'd have to be merged and released for the mne-bids test suite to pass, unfortunately. That's because we test mne-bids outputs against "stable" and "main" bids-validator.

I think one potential solution would be to:

1. expand these lines in the unit tests to also test fNIRS (stable, main, "fnirs branch"): https://github.com/mne-tools/mne-bids/blob/59ca433ad3e6ce87c120301e25e29c428df2e9d5/.github/workflows/unit_tests.yml#L112-L123
2.wrap fNIRS tests in mne-bids in a logic that only runs the tests when the bids-validator was installed from the fnirs branch

Then once the PR is merged, we can change this to only run fnirs tests on main ... and then finally after release, drop the conditional altogether


>  I don't know if the specification gets accepted, then the changes to the validator. Or if the validator gets merged, then the spec is submitted etc. Or if everything effectively happens at once. 

Typically we merge the spec first, and then the validator and examples --> ideally, it's all pretty much simultaneously. But I hope my plan from above works and we don't need to wait for any of this.

> Maybe @sappelhoff would like a chance to review?

I would be a blocker at this point (need to submit my dissertation by the 14th), so please don't wait for me

File: .github/workflows/unit_tests.yml (hunk 2 above)
Comment (sappelhoff) [reply .github/workflows/unit_tests.yml:178]: see: https://github.com/mne-tools/mne-bids/pull/406#issuecomment-1057762831

**Comment by hoechenberger** [comment]
Alright, CI should now run everything non-NIRS as it used to (with "vanilla" `bids-validator` stable and `main`); and NIRS tests should run on Windows, Linux, and macOS with Python 3.10 and @rob-luke's `bids-validator` branch.

**Comment by hoechenberger** [comment]
Codecov doesn't seem to be taking all tests into account, so the report isn't super useful …

But all integration tests passed successfully.

**Comment by hoechenberger** [comment]
I'll try to do a code review this afternoon and then it should be ready from my end.

@rob-luke I think what's still missing is an example for our gallery, https://mne.tools/mne-bids/dev/use.html#examples

**Comment by larsoner** [comment]
> 98.73% of diff hit

Nice -- Just need that example now @rob-luke !

**Comment by larsoner** [comment]
@hoechenberger I could quickly adapt any of these if they look reasonable to you:

https://mne.tools/mne-nirs/search.html?q=bids

**Comment by hoechenberger** [comment]
@larsoner We only need something that shows writing and reading, and then maybe some basic viz; and then we can say, for further info, check out the MNE-NIRS docs or something?

File: mne_bids/dig.py [hunk 3]
```diff
@@ -194,6 +196,62 @@ def _write_electrodes_tsv(raw, fname, datatype, overwrite=False):
...
+        Defaults to False.
+    verbose : bool
+        Set verbose output to True or False.
+    """
+    picks = _picks_to_idx(raw.info, 'fnirs', exclude=[], allow_empty=True)
+    sources = np.zeros(picks.shape)
+    detectors = np.zeros(picks.shape)
+    for ii in picks:
+        ch1_name_info = re.match(r'S(\d+)_D(\d+) (\d+)',
+                                 raw.info['chs'][ii]['ch_name'])
```
Comment (hoechenberger) [inline mne_bids/dig.py:219]: Can you please add a comment at to what's going on here?

File: mne_bids/write.py [hunk 4]
```diff
@@ -142,6 +145,32 @@ def _channels_tsv(raw, fname, overwrite=False):
...
+        ch_data["wavelength_nominal"] = [raw.info["chs"][i]["loc"][9] for i in
+                                         range(len(raw.ch_names))]
+
+        picks = _picks_to_idx(raw.info, 'fnirs', exclude=[], allow_empty=True)
+
+        sources = np.empty(picks.shape, dtype="<U20")
+        detectors = np.empty(picks.shape, dtype="<U20")
+        for ii in picks:
+            ch1_name_info = re.match(r'S(\d+)_D(\d+) (\d+)',
+                                     raw.info['chs'][ii]['ch_name'])
```
Comment (hoechenberger) [inline mne_bids/write.py:163]: Please add a comment what this is supposed to do

**Review by hoechenberger — CHANGES_REQUESTED** [review]
I left a few tiny comments / requests, otherwise the code looks good to me!

File: mne_bids/dig.py [hunk 5]
```diff
@@ -194,6 +196,62 @@ def _write_electrodes_tsv(raw, fname, datatype, overwrite=False):
...
+        xs[i] = raw.info["chs"][s_idx]["loc"][3]
+        ys[i] = raw.info["chs"][s_idx]["loc"][4]
+        zs[i] = raw.info["chs"][s_idx]["loc"][5]
+    for i, detector in enumerate(unique_detectors):
+        d_idx = np.where(detectors == detector)[0][0]
+        xs[i + n_sources] = raw.info["chs"][d_idx]["loc"][6]
+        ys[i + n_sources] = raw.info["chs"][d_idx]["loc"][7]
+        zs[i + n_sources] = raw.info["chs"][d_idx]["loc"][8]
+
+    ch_data = OrderedDict([
```
[the remark this answers is not in the document]
Comment (rob-luke) [reply mne_bids/dig.py]: There were other occurrences of OrderedDict, but I decided against doing a replace all in preference of doing that in a dedicated PR.

**Comment by rob-luke** [comment]
> Just need that example now

I'm on it. Just regenerating my example data in SNIRF to use as example input first.

File: examples/convert_nirs_to_bids.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+# Create a temporary directory to store the data in
+data_dir = tempfile.mkdtemp()
+file_path = f"{data_dir}/sub-01_task-tapping_nirs.snirf"
+
+# Download the data for subject 1.
+urllib.request.urlretrieve("https://github.com/rob-luke/BIDS-NIRS-Tapping/raw"
+                           "/01-Raw-to-SNIRF/sub-01/nirs/"
+                           "sub-01_task-tapping_nirs.snirf",
+                           file_path)
```
Comment (rob-luke) [inline examples/convert_nirs_to_bids.py:65]: Is there a cleaner way to do this? I don't know if this will work on all operating systems

**Review by rob-luke — COMMENTED** [review]
I have a minimal example running. But I did not have time to write the associated text yet, I will do that in the coming days. In the meantime, I had one question about how to download the data correctly.

File: examples/convert_nirs_to_bids.py (hunk 6 above)
Comment (hoechenberger) [reply examples/convert_nirs_to_bids.py:65]: MNE now officially depends on Pooch, so you could use that for downloading arbitrary files from arbitrary locations!

Check out this example here where they basically do what you want to do: download a file from GitHub:

https://www.fatiando.org/pooch/latest/retrieve.html#retrieve

Comment (hoechenberger) [reply examples/convert_nirs_to_bids.py:65]: I've implemented this now.

**Comment by hoechenberger** [comment]
@rob-luke Could you please make it such that the originally downloaded and the created BIDS files end up in `~/mne_data` somewhere (I think that's what we do in the other examples) so it'll be easier for users to actually find and browse those files?

File: examples/convert_nirs_to_bids.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+# For each subject, there were two baseline tasks (i) eyes open, (ii) eyes
+# closed, as well as four different motor imagery tasks.
+#
+# In this example, we will download the data for a single subject doing the
+# baseline task "eyes closed" and format it to the Brain Imaging Data Structure
+# (`BIDS <https://bids.neuroimaging.io/>`_).
+#
+
+# Create a temporary directory to store the data in
+data_dir = tempfile.mkdtemp()
```
Comment (larsoner) [inline examples/convert_nirs_to_bids.py:58]: Maybe it's easier for this example to depend on mne-nirs? Then you can just do `mne_nirs.datasets.fnirs_motor_group.data_path()`.

It only adds a dependency (plus its deps, like `nilearn`) to the doc build, not the library, so this seems preferable to me.

Comment (rob-luke) [reply examples/convert_nirs_to_bids.py:58]: Sure, will do now and we can revert if required.

**Comment by rob-luke** [comment]
Thanks for the feedback, you can see the latest render of the example at https://6055-89170358-gh.circle-artifacts.com/0/dev/auto_examples/convert_nirs_to_bids.html

File: examples/convert_nirs_to_bids.py (hunk 7 above)
Comment (hoechenberger) [reply examples/convert_nirs_to_bids.py]: Shouldn't we add MNE-NIRS to the library deps anyway?

Comment (hoechenberger) [reply examples/convert_nirs_to_bids.py]: Ah no, scratch that :)

**Review by hoechenberger — APPROVED** [review]
LGTM!

**Review by agramfort — APPROVED** [review]

**Comment by hoechenberger** [comment]
Thanks @rob-luke!

**Comment by rob-luke** [comment]
Thank you @hoechenberger!

**Pull request merged**